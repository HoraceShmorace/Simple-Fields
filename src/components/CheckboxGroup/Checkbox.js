import { useState } from 'react'
import { bool, func, oneOf, string } from 'prop-types'
import cx from 'classnames'
import styles from './Checkbox.module.css'

const Checkbox = ({ description, id, isDefaultChecked, labelFirst, label, onClick, type, value }) => {
  const [isChecked, setIsChecked] = useState(isDefaultChecked)
  const className = cx([
    styles.root,
    styles[type],
    { [styles.checked] : isChecked},
    labelFirst ? styles['label-before'] : styles['label-after']
  ])
  
  const toggleClick = () => {
    const nextIsChecked = !isChecked
    setIsChecked(nextIsChecked)
    if (typeof onClick === 'function') onClick(nextIsChecked)
  }

  return (
    <li className={className}>
      <input id={id} type="checkbox" value={value} defaultChecked={isChecked} onClick={toggleClick} />
      {
        labelFirst && 
        <label htmlFor={id}>
          {label}
        </label>
      }
      <button onClick={toggleClick}>
        <div className={styles.indicator} />
      </button>
      {
        !labelFirst && 
        <label htmlFor={id}>
          {label}
        </label>
      }
    </li>
  )
}

Checkbox.propTypes = {
  description: string,
  id: string.isRequired,
  isDefaultChecked: bool,
  labelFirst: bool,
  label: string,
  onClick: func,
  type: oneOf([
    'checkbox',
    'switch',
    'pills'
  ]),
  value: string
}

Checkbox.defaultProps = {
  description: null,
  isDefaultChecked: false,
  labelFirst: false,
  label: null,
  onClick: null,
  type: 'checkbox',
  value: ''
}

export default Checkbox
