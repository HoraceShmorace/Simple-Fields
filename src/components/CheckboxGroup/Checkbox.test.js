import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import Checkbox from './Checkbox'

describe('Checkbox', () => {
  describe('when rendered as CHECKED', () => {
    const props = {
      id: 'on',
      isCheckedByDefault: true,
      label: 'This is checked',
      onClick: jest.fn(),
      value: 'checkbox test'
    }

    const component = shallow(
      <Checkbox {...props} />
    )

    it('should match snapshot', () => {
      expect(toJson(component)).toMatchSnapshot()
    })
  })

  describe('when rendered as UNCHECKED', () => {
    const props = {
      id: 'off',
      isCheckedByDefault: false,
      label: 'This is not checked',
      onClick: jest.fn(),
      value: 'checkbox test'
    }

    const component = shallow(
      <Checkbox {...props} />
    )

    it('should match snapshot', () => {
      expect(toJson(component)).toMatchSnapshot()
    })
  })
})
