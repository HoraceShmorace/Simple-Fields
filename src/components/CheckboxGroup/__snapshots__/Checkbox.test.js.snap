// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox when rendered as CHECKED should match snapshot 1`] = `
<Fragment>
  <input
    className="fs-checkbox"
    defaultChecked={false}
    id="on"
    name="on"
    onClick={[MockFunction]}
    type="checkbox"
    value="checkbox test"
  />
  <label
    htmlFor="on"
  >
    This is checked
  </label>
</Fragment>
`;

exports[`Checkbox when rendered as UNCHECKED should match snapshot 1`] = `
<Fragment>
  <input
    className="fs-checkbox"
    defaultChecked={false}
    id="off"
    name="off"
    onClick={[MockFunction]}
    type="checkbox"
    value="checkbox test"
  />
  <label
    htmlFor="off"
  >
    This is not checked
  </label>
</Fragment>
`;
